### ISSUE TITLE
Doesn't support `#[no_std]`

### ISSUE THREAD

**Issue by japaric** [hashmismatch/packed_struct.rs#22]
### STR

``` console
$ cargo clone packed_struct && cd $_

$ # or any other no_std target
$ xargo check --target thumbv6m-none-eabi --no-default-features
   Compiling serde v1.0.27
   Compiling packed_struct v0.2.1 (file:///home/japaric/tmp/packed_struct)
   Compiling syn v0.11.11
error[E0463]: can't find crate for `std`
  |
  = note: the `thumbv6m-none-eabi` target may not be installed

error: aborting due to previous error

error: Could not compile `serde`.
warning: build failed, waiting for other jobs to finish...
error: build failed
```

The problem is that your serde dependency depends on `std` (by default).

I tried setting `default-features = false` for the serde dependency:

``` diff
 readme = "../README.md"

 [dependencies]
-serde = "1.0"
 serde_derive = "1.0"

+[dependencies.serde]
+default-features = false
+version = "1.0"
+
 [features]
 default = ["std"]
-std = []
+std = ["serde/std"]
 core_collections = []

 # comment this section when publishing new releases to crates.io!
-[dev-dependencies]
-packed_struct_codegen = "0.2"
+# [dev-dependencies]
+# packed_struct_codegen = "0.2"
```

I had to fix `src/internal_prelude/no_std.rs`

``` diff
 pub use core::marker::PhantomData;
 pub use core::iter;
 pub use core::cell::RefCell;
@@ -28,10 +27,8 @@ pub use collections::vec::Vec;
 pub use collections::string::*;
 #[cfg(feature="core_collections")]
 pub use collections::fmt::format as format_to_string;
-#[cfg(feature="core_collections")]
-pub use collections::fmt::{Display, Formatter};
+pub use core::fmt::{Display, Formatter};
 #[cfg(feature="core_collections")]
 pub use collections::borrow::Cow;
 #[cfg(feature="core_collections")]
 pub use collections::str::{from_utf8, FromStr};
```

and `src/packing.rs` as well

``` diff
     fn packed_bytes() -> usize;

     /// Pack the structure into a new byte vector.
+    #[cfg(any(feature="core_collections", feature="std"))]
     fn pack_to_vec(&self) -> Result<Vec<u8>, PackingError> {
         let mut buf = vec![0; Self::packed_bytes()];
         self.pack_to_slice(&mut buf)?;
```

and that made `packed_struct` compile with `--no-default-features` for the `no_std` target.

But the result was still unusable because it seems that `#[derive(PackedStruct)]` expands into some code that calls functions / methods inside the `debug_fmt` module and that module is not available on `no_std` because it requires dynamic memory allocation.

At this point I stopped looking further but could the generated code be tweaked to not require dynamic memory allocations when targeting `no_std`?

**Comment by rudib** [comment]
Yes, at the moment ```no_std``` isn't properly supported but that definitely is a priority for me. I suppose that there really should be three targets: Regular ```std```, ```no_std``` with the allocator (runtime debug formatting, etc...) and a non-allocating ```no_std``` target. The last one was never properly implemented as I wasn't interested in it.

I know that you are currently working with drivers and HALs, so my sample usage of this crate for sensor registers might spark an interest with you: https://github.com/hashmismatch/periphery.rs/blob/master/periphery_devices/invensense_mpu/registers.rs

I'm currently dealing with some family issues, but I might have some time for this in a couple of weeks.

**Comment by mcheshkov** [comment]
You could use https://github.com/kviver/bitfield-register, which supports `#[no_std]`, but is less feature-complete, and it's model of operation is different - each field access will do shift-mask stuff, but none required to wrap byte array with struct.

@rudib do you mind me sending a PR or two on this weekend? I'd love to see `#[no_std]` without alloc support. At first glance, codegen would need some refactoring, and `Vec` in debug print could be eliminated.

**Comment by rudib** [comment]
@mcheshkov I would be delighted. My main wish would be for a ```no_std``` CI environment that actually runs the unit tests, so that we can track if we break anything. My other project, [FreeRTOS.rs](https://github.com/hashmismatch/freertos.rs) has a build that actually sets up a QEMU emulator with an emulated STM32 MCU to test itself... Perhaps something similar could be built with a Rust-first approach, with no GCC in-between?

This codegen actually produced code that ran on a quadcopter flight controller board, but that was at least one year ago... It's really time-consuming to keep up with this stuff!

**Comment by japaric** [comment]
@mcheshkov Thanks for the suggestion. I ended up making a simple macro by example that looks like:

``` rust
// documentation is mandatory
register!(MACON3, 0, u8, {
    #[doc = "MAC will operate in Full-Duplex mode"]
    fuldpx @ 0,
    #[doc = "The type/length field of transmitted and received frames will be checked"]
    frmlnen @ 1,
    #[doc = "Frames bigger than MAMXFL will be aborted when transmitted or received"]
    hfrmen @ 2,
    #[doc = "No proprietary header is present"]
    phdren @ 3,
    #[doc = "MAC will append a valid CRC to all frames transmitted regardless of PADCFG bit"]
    txcrcen @ 4,
    #[doc = "All short frames will be zero-padded to 64 bytes and a valid CRC will then be appended"]
    padcfg @ 5..7,
});
```

and that's used like this:

``` rust
        self.write_control_register(
            bank2::Register::MACON3,
            bank2::MACON3::default() // <-
                .frmlnen(1)
                .txcrcen(1)
                .padcfg(0b001)
                .bits(),
        )?;
```

This is for the implementation / private API of a driver I'm working on so I can always change it later. It doesn't support enums but I don't really need them at the moment so this will make do for now.

**Comment by rudib** [comment]
Just released 0.2.2 which should work with ```no_std```. The ```packed_struct_nostd_tests``` successfully passes a xargo build, as per the original issue.

Additional support for formatting and using the ```alloc``` crate is still missing, as it looks like its not trivial to setup a CI environment for this at the moment. Suggestions are welcome.
